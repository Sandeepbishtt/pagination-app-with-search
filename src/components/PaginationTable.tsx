import { useSelector } from "react-redux";
import { useState, ChangeEvent } from "react";
import { getAllData } from "../Redux/PostSlice";
import { Button, Paper } from "@material-ui/core";
import { CircularProgress } from "@material-ui/core";

import moment from 'moment'
import {
  InputBase,
  TableContainer,
  Table,
  TableHead,
  TableRow,
  TableCell,
  TableBody,
} from "@material-ui/core";
import { Pagination } from "@mui/material";
import Detail from "./ModalJson";
const classes = {
  formStyle: {
    margin: "1rem",
    display: "flex",
    justifyContent: "end",
    alignItems: "center",
  },
  loading :{
     marginLeft: "50%", marginTop: "15rem" 
  },
  paginate :{ margin: "2rem", marginLeft: "50%" }

};

const PaginationTable = (props: any) => {
  const data = useSelector(getAllData);
  const [confirmOpen, setConfirmOpen] = useState(false);
  const [dataForModal, setDataForModal]: any[] = useState([]);
console.log(data)

  const [page, setPage] = useState(0);
  const [pageValue, setPageValue] = useState(0);

  const modalClickHandler = (id: number, val: any) => {
    setConfirmOpen(true);
    setDataForModal(val);
  };

  const [input, setInput] = useState("");
  const [showSearch, setShowSearch] = useState(true);
  const [filterData, setFilterData] = useState([]);
  const [pageCountForSearch, setPageCountForSearch] = useState(true);

  
  const changeHandler = (e: any, val: any) => {
    setPage(val);
    setPageValue(val - 1);
  };



  const searchHandler = (e: any) => {
    setPageCountForSearch(false);
    e.preventDefault();
    const dataBySearch = data.slice(pageValue * 20, pageValue * 20 + 20).filter((val: any) => {
      return (
        val.title.toLowerCase().trim().includes(input.toLowerCase().trim()) ||
        moment(val.created_at).format('LLLL').toLowerCase().trim().includes(input.toLowerCase().trim())
      );
    });
    setShowSearch(false);
    setFilterData(dataBySearch);
    setInput("");
   
  };
  const refreshHandler = () =>{
     if (input === "") {
      setFilterData([]);
      setShowSearch(true);
      setPageCountForSearch(true);
    }
  }


  return (
    <>
      <Detail
        data={dataForModal}
        title="Post RawJSON"
        open={confirmOpen}
        setOpen={setConfirmOpen}
      />
 {props.showLoad ? <CircularProgress size ={80}  style={classes.loading}/> :
      <div style={{ height: 750 }}>
        <form style={classes.formStyle} onSubmit={searchHandler}>
          <InputBase
            style={{ border: "2px solid grey" }}
            placeholder="Search"
            type="text"
            value={input}
            required
            onChange={(event: ChangeEvent<HTMLInputElement>) =>
              setInput(event.target.value)
            }
          ></InputBase>
         {showSearch ? <Button onClick={searchHandler} variant='contained' color='primary' disabled = {input === ''} >Search </Button> : <Button variant='contained' color='primary' onClick={refreshHandler} >Back</Button>}
        </form>
        <Paper>
          <TableContainer>
            <Table stickyHeader aria-label="sticky table">
              <TableHead>
                <TableRow>
                 <TableCell>Created_at</TableCell>
                 <TableCell>Title</TableCell>
                 <TableCell>Author</TableCell>
                 <TableCell>URL</TableCell>
               
                </TableRow>
              </TableHead>
              <TableBody>
                {showSearch
                  ? data
                      .slice(pageValue * 20, pageValue * 20 + 20)
                      .map((val: any, index: any) => {
                        return (
                          <TableRow
                            key={val.created_at_i}
                            onClick={() =>
                              modalClickHandler(val.created_at_i, val)
                            }
                          >
                            <TableCell>{moment(val.created_at).format('LLLL')}</TableCell>
                            <TableCell>{val.title}</TableCell>
                            <TableCell>{val.author}</TableCell>
                            <TableCell>{val.url} </TableCell>
                          </TableRow>
                        );
                      })
                  : filterData
                      .map((val: any, index: any) => {
                        return (
                          <TableRow
                            key={val.created_at_i}
                            onClick={() =>
                              modalClickHandler(val.created_at_i, val)
                            }
                          >
                            <TableCell>{ moment(val.created_at).format('LLLL')}</TableCell>
                            <TableCell>{val.title}</TableCell>
                            <TableCell>{val.author}</TableCell>
                            <TableCell>{val.url} </TableCell>
                          </TableRow>
                        );
                      })}
              </TableBody>
            </Table>

            {pageCountForSearch && (
              <Pagination
                style={classes.paginate}
                count={ props.pageNumber + 1}
                page={page}
                onChange={changeHandler}
                variant="outlined"
                shape="rounded"
                color="primary"
              />
            )}
          </TableContainer>
        </Paper>
      </div>
}
    </>
  );
};

export default PaginationTable;
