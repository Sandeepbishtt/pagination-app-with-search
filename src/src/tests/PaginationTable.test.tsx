import React from 'react';
import { render, screen } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import {Provider} from 'react-redux'
import store from '../Redux/Store'
import PaginationTable from '../components/PaginationTable';
import '@testing-library/jest-dom'
import ReactDOM from 'react-dom'
test('Example', async () => {
    render( <BrowserRouter  >
        <Provider store={store} >
        <PaginationTable />
          </Provider>
          </BrowserRouter>);
    const text = await screen.findByText(/Created_at/i)
    expect(text).toBeInTheDocument()
  }) 

  describe("Home Component tests", () => {
    let container: HTMLDivElement;
    beforeEach(() => {
      container = document.createElement("div");
      document.body.appendChild(container);
      ReactDOM.render(
        <React.StrictMode>
          <BrowserRouter>
            <Provider store={store}>
              <PaginationTable />
            </Provider>
          </BrowserRouter>
        </React.StrictMode>,
        container
      );
    });
    afterEach(() => {
      document.body.removeChild(container);
      container.remove();
    });
    it("Render correctly initial document", () => {
      expect(container).toMatchSnapshot();
    });

    it('checking input in Component', () => {
      const inputs = container.querySelectorAll('input');
      expect(inputs).toHaveLength(1);
  })
  it('checking button in component', () => {
      const buttons = container.querySelectorAll('button');
      expect(buttons).toHaveLength(3);
  })


  });
  