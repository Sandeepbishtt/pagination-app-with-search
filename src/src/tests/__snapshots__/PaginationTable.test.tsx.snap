// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component tests Render correctly initial document 1`] = `
<div>
  <div
    style="height: 750px;"
  >
    <form
      style="margin: 1rem; display: flex; justify-content: end; align-items: center;"
    >
      <div
        class="MuiInputBase-root"
        required=""
        style="border: 2px solid grey;"
      >
        <input
          class="MuiInputBase-input"
          placeholder="Search"
          required=""
          type="text"
          value=""
        />
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Search 
        </span>
      </button>
    </form>
    <div
      class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiTableContainer-root"
      >
        <table
          aria-label="sticky table"
          class="MuiTable-root MuiTable-stickyHeader"
        >
          <thead
            class="MuiTableHead-root"
          >
            <tr
              class="MuiTableRow-root MuiTableRow-head"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
                scope="col"
              >
                Created_at
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
                scope="col"
              >
                Title
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
                scope="col"
              >
                Author
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
                scope="col"
              >
                URL
              </th>
            </tr>
          </thead>
          <tbody
            class="MuiTableBody-root"
          />
        </table>
        <nav
          aria-label="pagination navigation"
          class="MuiPagination-root MuiPagination-outlined css-1oj2twp-MuiPagination-root"
          style="margin: 2rem 2rem 2rem 50%;"
        >
          <ul
            class="MuiPagination-ul css-wjh20t-MuiPagination-ul"
          >
            <li>
              <button
                aria-label="Go to previous page"
                class="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-outlined MuiPaginationItem-rounded MuiPaginationItem-outlinedPrimary Mui-disabled MuiPaginationItem-previousNext css-r93niq-MuiButtonBase-root-MuiPaginationItem-root"
                disabled=""
                tabindex="-1"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
                  data-testid="NavigateBeforeIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                  />
                </svg>
              </button>
            </li>
            <li>
              <button
                aria-label="Go to next page"
                class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-outlined MuiPaginationItem-rounded MuiPaginationItem-outlinedPrimary MuiPaginationItem-previousNext css-r93niq-MuiButtonBase-root-MuiPaginationItem-root"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
                  data-testid="NavigateNextIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                  />
                </svg>
              </button>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </div>
</div>
`;
